### PR TITLE
Adds API endpoints for users and research groups

### PR DESCRIPTION
**Pull request by Comeani**
This PR builds out a mechanism for doing a ResearchGroup sync with Slurm in a similar fashion to how the Users are synced to LDAP. 

It introduces some more functions for getting specific info from Slurm like PI user name (We use an account's "Description" field to denote PI username, and could do a bulk reassignment to the potentially more fitting ["Coordinators" field](https://slurm.schedmd.com/sacctmgr.html#SECTION_LIST/SHOW-ACCOUNT-FORMAT-OPTIONS), need to look into it more), and the usernames of users under an account. 

With these, changes to `ResearchGroup.pi` or `ResearchGroup.members` for existing groups can be brought into keystone and, for new groups, a new ResearchGroup can be established.

It does not include a Celery task schedule item yet as I was not sure what the refresh interval should be relative to the hourly update of users.

### CODE REVIEW

**Comment by Comeani** [comment]
@djperrefort I think this is ready for review, but before that, did I go in the right direction with this? I mostly tried to emulate how you did it with users.

**Comment by Comeani** [comment]
We are currently using a Slurm Account's "description" field to house the PI username upon creation of the account. The documentation describes this as "an arbitrary string describing an account".

[nlc60@login0 ~] : sacctmgr show account where account=sam
   Account                Descr                  Org 
---------- -------------------- -------------------- 
       sam             bmooreii center for research+ 

At first glance, it does not seem like other Slurm shop HPC groups would use this for this purpose. Slurm does support the notion of a "coordinator" role for one or more users under a Slurm account with a combination of special privileges we may or may not want PIs to have:

    A special privileged user, usually an account manager, that can add users or sub-accounts to the account they are coordinator over. This should be a trusted person since they can change limits on account and user associations, as well as cancel, requeue or reassign accounts of jobs inside their realm.

**Review by djperrefort — COMMENTED** [review]
Structurally this looks solid. However, I'm still not sure if regularly syncing against slurm is the way we want to go.

In my head clusters are separate, unrelated entities in keystone. This means we could theoretically administrate two separate clusters which may have slurm accounts with different user associations. Maybe it's worth moving away from that mental model. Either way, I think we should treat research groups primarily as a permissions mechanism to determine what users can and can't do within keystone. I want to be cautious we don't try to leverage keystone as an admin layer over slurm.

**Comment by Comeani** [comment]
@djperrefort This PR now represents the changes to the application required to enable the logic in [this research group ingest script in crc-ops](https://github.com/pitt-crc/crc-ops/pull/14/files#diff-a88072d97e68ae0bee393758bd035311f826c4924467a75124d31e5ec7855f15).

I'm realizing I may have slightly duplicated some of the effort that was otherwise going into creating the users endpoint. The script mentioned above only needs to GET from keystone's `User` list (to compare with Slurm user names) and PUT new `ResearchGroup` entries in using the `id`s of the `User`s, so the views represent that.

File: keystone_api/apps/users/views.py
Comment (djperrefort) [inline keystone_api/apps/users/views.py:42]: This can be lessened to authenticated users. Eventually we will want to support users searching for other users and so that teams can add members to research groups.

File: keystone_api/main/settings.py
Comment (djperrefort) [inline keystone_api/main/settings.py:150]: Expose this as an environmental variable and update the readme with a description for the new setting value.

File: keystone_api/main/settings.py
Comment (djperrefort) [inline keystone_api/main/settings.py:153]: This `if DEBUG` block was removed in earlier work. It looks like it was added back in here by accident.

File: keystone_api/plugins/slurm.py
Comment (djperrefort) [inline keystone_api/plugins/slurm.py:47]: Remove extra whitespace

File: keystone_api/plugins/slurm.py
Comment (djperrefort) [inline keystone_api/plugins/slurm.py:61]: Add whitespace between docs and code

File: keystone_api/plugins/slurm.py
Comment (djperrefort) [inline keystone_api/plugins/slurm.py:80]: remove extra whitespace

File: keystone_api/apps/users/views.py
Comment (djperrefort) [inline keystone_api/apps/users/views.py:26]: The `ModelViewSet` base class already provides create functionality ([docs](https://www.django-rest-framework.org/api-guide/viewsets/#modelviewset)). Is there a reason it is overwritten here?

**Review by djperrefort — CHANGES_REQUESTED** [review]
This looks great. I have a few small changes/questions in he comments. One larger request is that function tests need to be added in the `keystone_api/tests` directory - one test file per endpoint. These tests should focus on user permissions and can be mostly copy pasted from other examples you'll find in that directory.

File: keystone_api/apps/users/views.py
Comment (Comeani) [reply keystone_api/apps/users/views.py:26]: Honestly I was just trying not to stray from how I saw it being set up for the `AllocationRequestReviewViewSet`. I can see why there was an override there now.

File: keystone_api/main/settings.py
Comment (Comeani) [reply keystone_api/main/settings.py:150]: I think I set this more because it felt like the token was expiring while I was tinkering than because it really needs to be an hour. I don't have a great indication yet towards how long pulling the ResearchGroup info in from Slurm will take out on the cluster, so I'll just pull it for now.

**Comment by Comeani** [comment]
@djperrefort I'm realizing having the tests designed and working as expected out the gate for these endpoints is going to be easier with the permissions work you started under the [users_api](https://github.com/pitt-crc/keystone-api/tree/users_api) branch in place. 

Was that far enough along that I could work on finishing what you were setting up instead of trying to do the tests here (with the scope of this PR going back to what it was before your review, and having it wait until the changes in users_api are merged in) or would it be better to copy the code relevant to the views/permissions to this branch?

**Comment by djperrefort** [comment]
The user's api branch tries to implement a much more complicated permissions model where users can self administrate and research group admins can modify group membership. I should have deleted it a while ago (and will cean up my old branches right after this comment). For now, that work is out of scope for this PR. Those features won't be needed for a long while - certainly not before we have a front end.

Lets merge the PR as is and call it a day.

**Review by djperrefort — APPROVED** [review]

**Comment by Comeani** [comment]
@djperrefort Before I merge this in, would it be worth committing the tests in their current state (passing but probably not with the statuses we'd prefer for once we open the endpoint access to users) or should I just keep them in a separate branch for now kinda like we did with the Slurm API logic?

**Comment by djperrefort** [comment]
@Comeani If you're happy with the manual testing you've done then I say merge away 👍

**Pull request merged**
